Compilation errors in auto-generated Topology.cpp

| | |
|:---|:---|
|**_F´ Version_**| fprime-devel branch cloned 3/11/2025 |
|**_Affected Component_**|   |
---
## Problem Description

The "-pedantic" option has been added to the  options passed to c++ which causes the auto-generated <deployment>Topology.cpp (e.g. MyDeployment/Top/MyDeploymentTopology.cpp) file to fail to compile with the below warnings ( `fprime-util build -v` was used to provide additional info)

```
cd /home/USER/dev/fprime/build-fprime-automatic-native/MyDeployment/Top && /usr/bin/c++ -DFW_BAREMETAL_SCHEDULER=0 -DTGT_OS_TYPE_LINUX -I/home/USER/dev/fprime -I/home/USER/dev/fprime/config -I/home/USER/dev/fprime/build-fprime-automatic-native -I/home/USER/dev/fprime/build-fprime-automatic-native/F-Prime -I/home/USER/dev/fprime/build-fprime-automatic-native/config -isystem /home/USER/dev/fprime/cmake/platform/types -std=c++11 -Wall -Wextra -Werror -Wold-style-cast -Wno-unused-parameter -Wshadow -pedantic -Wconversion -Wno-format-truncation -DASSERT_FILE_ID=0xa9eade69 -DASSERT_RELATIVE_PATH='"MyDeployment/Top/MyDeploymentTopology.cpp"' -MD -MT MyDeployment/Top/CMakeFiles/MyDeployment_Top.dir/MyDeploymentTopology.cpp.o -MF CMakeFiles/MyDeployment_Top.dir/MyDeploymentTopology.cpp.o.d -o CMakeFiles/MyDeployment_Top.dir/MyDeploymentTopology.cpp.o -c /home/USER/dev/fprime/MyDeployment/Top/MyDeploymentTopology.cpp
/home/USER/dev/fprime/MyDeployment/Top/MyDeploymentTopology.cpp: In function ‘void configureTopology(const MyDeployment::TopologyState&)’:
/home/USER/dev/fprime/MyDeployment/Top/MyDeploymentTopology.cpp:125:38: error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]
  125 |     configurationTable.entries[0] = {.depth = 100, .priority = 0};
      |                                      ^
/home/USER/dev/fprime/MyDeployment/Top/MyDeploymentTopology.cpp:125:52: error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]
  125 |     configurationTable.entries[0] = {.depth = 100, .priority = 0};
      |                                                    ^
/home/USER/dev/fprime/MyDeployment/Top/MyDeploymentTopology.cpp:127:38: error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]
  127 |     configurationTable.entries[1] = {.depth = 500, .priority = 2};
      |                                      ^
/home/USER/dev/fprime/MyDeployment/Top/MyDeploymentTopology.cpp:127:52: error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]
  127 |     configurationTable.entries[1] = {.depth = 500, .priority = 2};
      |                                                    ^
/home/USER/dev/fprime/MyDeployment/Top/MyDeploymentTopology.cpp:129:38: error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]
  129 |     configurationTable.entries[2] = {.depth = 100, .priority = 1};
      |                                      ^
/home/USER/dev/fprime/MyDeployment/Top/MyDeploymentTopology.cpp:129:52: error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]
  129 |     configurationTable.entries[2] = {.depth = 100, .priority = 1};
```


## Context / Environment

Execute `fprime-util version-check` and share the output.

```
:~/dev/fprime$ fprime-util version-check
Operating System: Linux
CPU Architecture: x86_64
Platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.39
Python version: 3.12.3
CMake version: 3.28.3
Pip version: 25.0.1
Pip packages:
    fprime-tools==3.6.1
    fprime-gds==3.6.1
    fprime-fpp-*==2.3.0
Project submodules:
```

## How to Reproduce

1. Checkout a clone of the devel branch from the fprime repo ("fprime-bootstrap clone --rename issue-repro https://github.com/nasa/fprime.git")
2. activate the venv (`. fprime-venv/bin/activate`)
3. generate & build the project (`fprime-util generate && fprime-util build`)
4. add a deployment (`fprime-util new --deployment`, hit <ENTER> to accept defaults for all prompts)
5. build the project (`fprime-util build -v`) FAILS
```
/home/USER/dev/issue-repro/MyDeployment/Top/MyDeploymentTopology.cpp: In function ‘void configureTopology(const MyDeployment::TopologyState&)’:
/home/USER/dev/issue-repro/MyDeployment/Top/MyDeploymentTopology.cpp:125:38: error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]
  125 |     configurationTable.entries[0] = {.depth = 100, .priority = 0};
      |                                      ^
/home/USER/dev/issue-repro/MyDeployment/Top/MyDeploymentTopology.cpp:125:52: error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]
  125 |     configurationTable.entries[0] = {.depth = 100, .priority = 0};
      |                                                    ^
/home/USER/dev/issue-repro/MyDeployment/Top/MyDeploymentTopology.cpp:127:38: error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]
  127 |     configurationTable.entries[1] = {.depth = 500, .priority = 2};
      |                                      ^
/home/USER/dev/issue-repro/MyDeployment/Top/MyDeploymentTopology.cpp:127:52: error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]
  127 |     configurationTable.entries[1] = {.depth = 500, .priority = 2};
      |                                                    ^
/home/USER/dev/issue-repro/MyDeployment/Top/MyDeploymentTopology.cpp:129:38: error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]
  129 |     configurationTable.entries[2] = {.depth = 100, .priority = 1};
      |                                      ^
/home/USER/dev/issue-repro/MyDeployment/Top/MyDeploymentTopology.cpp:129:52: error: C++ designated initializers only available with ‘-std=c++20’ or ‘-std=gnu++20’ [-Werror=c++20-extensions]
  129 |     configurationTable.entries[2] = {.depth = 100, .priority = 1};
  ```

## Expected Behavior

The auto-generated code should compile without any failures

--- a/.github/workflows/integration-tests.yml
+++ b/.github/workflows/integration-tests.yml
@@ -26,20 +26,16 @@ jobs:
         uses: actions/setup-python@v4
         with:
           python-version: ${{ matrix.python-version }}
-      - name: Install dependencies
+      - name: Install System dependencies
         run: |
           python -m pip install --upgrade pip urllib3 setuptools_scm
           sudo apt-get install ninja-build
-      - name: clone fprime and gps-tutorial
+      - name: Clone fprime and install requirements.txt
         run: |
           git clone --recurse-submodules -b ${{ matrix.fprime-version }} https://github.com/nasa/fprime.git
-          git clone --recurse-submodules https://github.com/fprime-community/gps-tutorial.git
           pip install -r fprime/requirements.txt || true # Attempt to install fprime package when available
-      - name: Generate correct fprime version in gps-tutorial
-        working-directory: gps-tutorial
-        run: git submodule update --recursive
-      - name: Install Current fpp-tools
-        run: pip install .
+      - name: Install Current fprime-tools
+        run: pip install -U .
       # 0x72ad3277 is Fw/Types/Assert.cpp, a file that should be consistent across releases
       - name: Test generation and building on Ref
         working-directory: fprime/Ref
@@ -47,16 +43,8 @@ jobs:
           fprime-util generate
           fprime-util build
           fprime-util hash-to-file 0x72ad3277
-      - name: Test generation and building on gps-tutorial
-        working-directory: gps-tutorial/GpsApp
-        run: |
-          fprime-util generate
-          fprime-util build
-          fprime-util hash-to-file 0x72ad3277
       - name: Test generation and building on Ref with Ninja
         working-directory: fprime/Ref
-        # There is a bug in v3.0.0 that causes ninja builds to fail
-        if: matrix.fprime-version != 'v3.0.0'
         run: |
           fprime-util generate --ninja --build-cache ./build-ninja
           fprime-util build --build-cache ./build-ninja

--- a/src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/{{cookiecutter.deployment_name}}/Top/{{cookiecutter.deployment_name}}Topology.cpp
+++ b/src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/{{cookiecutter.deployment_name}}/Top/{{cookiecutter.deployment_name}}Topology.cpp
@@ -123,12 +123,16 @@ void configureTopology(const TopologyState& state) {
     // Note: Uncomment when using Svc:TlmPacketizer
     // tlmSend.setPacketList({{cookiecutter.deployment_name}}PacketsPkts, {{cookiecutter.deployment_name}}PacketsIgnore, 1);
 
+    // ComQueue configuration
     // Events (highest-priority)
-    configurationTable.entries[0] = {.depth = 100, .priority = 0};
+    configurationTable.entries[0].depth = 100;
+    configurationTable.entries[0].priority = 0;
     // Telemetry
-    configurationTable.entries[1] = {.depth = 500, .priority = 2};
+    configurationTable.entries[1].depth = 500;
+    configurationTable.entries[1].priority = 2;
     // File Downlink
-    configurationTable.entries[2] = {.depth = 100, .priority = 1};
+    configurationTable.entries[2].depth = 100;
+    configurationTable.entries[2].priority = 1;
     // Allocation identifier is 0 as the MallocAllocator discards it
     comQueue.configure(configurationTable, 0, mallocator);
 {%- if (cookiecutter.com_driver_type in ["TcpServer", "TcpClient"]) %}

